libfreerdp/core/nla.c

```diff
@@ -526,41 +536,228 @@ void nla_identity_free(SEC_WINNT_AUTH_IDENTITY* identity)
+		{
+			WLog_ERR(TAG, "error setting X509 anchors");
+			return FALSE;
+		}
+	}
+
+	if (!smartcard_enumerateCerts(settings, certs, MAX_SC_CERTS, &count))
+	{
+		WLog_ERR(TAG, "unable to list smartcard certificates");
+		return FALSE;
+	}
+
+	if (!count)
+	{
+		WLog_ERR(TAG, "no smartcard certificates found");
+		return FALSE;
+	}
+
+	if (count != 1)
+		goto setup_pin;
+
+	/*
+	 * just one result let's try to fill missing parameters
+	 */
+	if (!settings->Username && certs[0].userHint)
+	{
+		settings->Username = strdup(certs[0].userHint);
+		if (!settings->Username)
+		{
+			WLog_ERR(TAG, "unable to copy certificate username");
+			goto out;
+		}
+	}
+
+	if (!settings->Domain && certs[0].domainHint)
+	{
+		settings->Domain = strdup(certs[0].domainHint);
+		if (!settings->Domain)
+		{
+			WLog_ERR(TAG, "unable to copy certificate domain");
+			goto out;
+		}
+	}
+
+	if (!settings->ReaderName && certs[0].reader)
+	{
+		if (ConvertFromUnicode(CP_UTF8, 0, certs[0].reader, -1, &settings->ReaderName, 0, NULL,
+		                       NULL) < 0)
+		{
+			WLog_ERR(TAG, "unable to copy reader name");
+			goto out;
+		}
+	}
+
+	if (!settings->ContainerName && certs[0].containerName)
+	{
+		settings->ContainerName = strdup(certs[0].containerName);
+		if (!settings->ContainerName)
+		{
+			WLog_ERR(TAG, "unable to copy container name");
+			goto out;
+		}
+	}
+
+	memcpy(nla->kerberosSettings.certSha1, certs[0].sha1Hash,
+	       sizeof(nla->kerberosSettings.certSha1));
+
+#ifndef _WIN32
+	if (certs[0].pkinitArgs)
+	{
+		kerbSettings->pkinitX509Identity = strdup(certs[0].pkinitArgs);
+		if (!kerbSettings->pkinitX509Identity)
+		{
+			WLog_ERR(TAG, "unable to copy pkinitArgs");
+			goto out;
+		}
+	}
+#endif /* _WIN32 */
+
+setup_pin:
+	if (!settings->Pin)
+	{
+		if (settings->SmartcardEmulation)
+		{
+			const char* dupSrc = settings->SmartcardPin;
+			if (!dupSrc)
+				dupSrc = "";
+
+			settings->Pin = strdup(dupSrc);
+			if (!settings->Pin)
+			{
+				WLog_ERR(TAG, "error setting pin");
+				return FALSE;
+			}
+		}
+		else
+		{
+			freerdp* instance = nla->instance;
+			if (!instance->AuthenticateEx)
+			{
+				WLog_ERR(TAG, "no pin configured and no instance->AuthenticateEx callback");
+				return TRUE;
+			}
+
+			if (!instance->AuthenticateEx(instance, &settings->Username, &settings->Pin,
+			                              &settings->Domain, AUTH_SMARTCARD_PIN))
+			{
+				WLog_ERR(TAG, "no pin code entered");
+				return TRUE;
+			}
+		}
+	}
+
+	ret = TRUE;
+out:
+	for (i = 0; i < count; i++)
+		smartcardCert_Free(&certs[i]);
+
+	return ret;
+}
+
+static BOOL nla_client_setup_identity(rdpNla* nla)
+{
 	WINPR_SAM* sam;
 	WINPR_SAM_ENTRY* entry;
-	nla_set_state(nla, NLA_STATE_INITIAL);
+	BOOL PromptPassword = FALSE;
+	rdpSettings* settings = nla->settings;
+	freerdp* instance = nla->instance;
 
-	if (settings->RestrictedAdminModeRequired)
-		settings->DisableCredentialsDelegation = TRUE;
+	if (settings->SmartcardLogon)
+	{
+#ifdef _WIN32
```

Comment:
This should be a check for `UNICODE` or am I wrong?

Reply:
Also, since we use `utf8` for most internal stuff, wouldn´t it be better to always use the `CredMarshalCredentialW` on non windows?

Reply:
in fact for non windows we don't use `CredMarshalCredential` at all (I failed to find a way to emulate that in winPR), but you're right it should be the CredMarshalCredentialA version (as we call AcquireA functions after that in nla)

Reply:
what is the problem with a custom `CredMashalCredential` function that appends a hash to a username with some defined delimiter?
Since we control the SSPI below we can use the appropriate `CredUnmarshalCredential` (is that named like this?) to extract the data as required again.